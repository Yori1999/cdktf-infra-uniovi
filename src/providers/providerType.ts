export enum ProviderType {
  DOCKER = "docker",
  AWS = "aws",
}
