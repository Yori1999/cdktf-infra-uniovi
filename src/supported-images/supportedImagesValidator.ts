export function validateImageExists(imageIdentifier: string): void {
  if (!imageIdentifier || imageIdentifier.trim().length === 0) {
    throw new Error(
      "This version of the machine doesn't exist for the given provider",
    );
  }
}
