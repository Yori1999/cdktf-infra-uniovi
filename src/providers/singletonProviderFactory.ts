import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { DockerProvider } from "@cdktf/provider-docker/lib/provider";
import { TerraformProvider } from "cdktf";
import { Construct } from "constructs";
import { ProviderType } from "./providerType";

export class SingletonProviderFactory {
  // We could've chosen not to return anything but... we might need it
  static getProvider(
    providerType: ProviderType,
    scope: Construct,
    existingProvider?: TerraformProvider,
  ): TerraformProvider {
    if (existingProvider) {
      return existingProvider;
    }

    if (!this.providers[providerType]) {
      const provider = SingletonProviderFactory.createProvider(
        providerType,
        scope,
      );
      this.providers[providerType] = provider;
    }

    return this.providers[providerType];
  }

  private static providers: { [key in ProviderType]?: TerraformProvider } = {};

  /**
   * Creates a provider based on a given type, with either the default properties or the properties passed by the user
   * @param providerType the type of the provider to create
   * @param scope
   * @param properties a set of properties that may override the default ones
   * @returns
   */
  private static createProvider(
    providerType: ProviderType,
    scope: Construct,
  ): TerraformProvider {
    switch (providerType) {
      case ProviderType.DOCKER:
        return new DockerProvider(scope, "docker", {});
      case ProviderType.AWS:
        return new AwsProvider(scope, "aws", {
          region: "eu-west-1",
        });
    }
  }

  // Prevents instantiation, just in case
  private constructor() {}
}
