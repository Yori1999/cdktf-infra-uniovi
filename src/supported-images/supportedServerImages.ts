import { DebianVersion, supportedDebianImages } from "./supportedMachineImages";
import { ProviderType } from "../providers/providerType";

export enum NginxVersion {
  LATEST = "latest",
}
export const supportedNginxImages = Object.freeze({
  [ProviderType.DOCKER]: {
    [NginxVersion.LATEST]: "nginx:latest",
  },
  [ProviderType.AWS]: {
    [NginxVersion.LATEST]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
  },
}) satisfies Record<ProviderType, Record<NginxVersion, string>>;
export const supportedHardenedNginxImages = Object.freeze({
  [ProviderType.DOCKER]: {
    [NginxVersion.LATEST]: "rapidfort/nginx-official:latest",
  },
  [ProviderType.AWS]: {
    [NginxVersion.LATEST]: "ami-04ca7bd38fd01789c", // NGINX Plus App Protect - Ubuntu 20.04 Developer AMI
  },
}) satisfies Record<ProviderType, Record<NginxVersion, string>>;

export enum ApacheVersion {
  LATEST = "latest",
  APACHE_DEBIAN = "apacheDebian",
  APACHE_ALPINE = "apacheAlpine",
}
export const supportedApacheImages = Object.freeze({
  [ProviderType.DOCKER]: {
    [ApacheVersion.LATEST]: "httpd:latest",
    [ApacheVersion.APACHE_DEBIAN]: "httpd:bookworm",
    [ApacheVersion.APACHE_ALPINE]: "httpd:alpine",
  },
  [ProviderType.AWS]: {
    [ApacheVersion.LATEST]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
    [ApacheVersion.APACHE_DEBIAN]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
    [ApacheVersion.APACHE_ALPINE]: "", // doesn't exist, so we cannot assign this version but we need to include it
  },
}) satisfies Record<ProviderType, Record<ApacheVersion, string>>;
export const supportedHardenedApacheImages = Object.freeze({
  [ProviderType.DOCKER]: {
    [ApacheVersion.LATEST]: "rapidfort/apache-official:latest", // It is equivalent to httpd:latest
    [ApacheVersion.APACHE_DEBIAN]: "rapidfort/apache-official:2.4.63-bookworm",
    [ApacheVersion.APACHE_ALPINE]:
      "rapidfort/apache-official:2.4.63-alpine3.22",
  },
  [ProviderType.AWS]: {
    [ApacheVersion.LATEST]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
    [ApacheVersion.APACHE_DEBIAN]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
    [ApacheVersion.APACHE_ALPINE]: "", // doesn't exist, so we cannot assign this version but we need to include it
  },
}) satisfies Record<ProviderType, Record<ApacheVersion, string>>;
