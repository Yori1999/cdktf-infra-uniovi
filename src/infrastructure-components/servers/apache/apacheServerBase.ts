import { TerraformProvider } from "cdktf";
import { Construct } from "constructs";
import {
  InternalMachineComponentPropsInterface,
  ServerPropsInterface,
} from "../../../props/props";
import { IDeployStrategy } from "../../../providers/providerDeployStrategy/deployStrategy";
import { ProviderDeployStrategyFactory } from "../../../providers/providerDeployStrategy/providerDeployStrategyFactory";
import { ProviderType } from "../../../providers/providerType";
import { SingletonProviderFactory } from "../../../providers/singletonProviderFactory";
import { validateImageExists } from "../../../supported-images/supportedImagesValidator";
import { ApacheVersion } from "../../../supported-images/supportedServerImages";
import { checkIsValidId } from "../../../utils/stringUtils";

export abstract class ApacheServerBase extends Construct {
  protected createdApacheServer?: Construct;

  protected abstract get supportedApacheImagesMap(): Record<
    string,
    Record<string, string>
  >; // otherwise, JSII throws error JSII1003: Only string-indexed map types are supported

  constructor(
    scope: Construct,
    id: string,
    version: ApacheVersion,
    serverProps: ServerPropsInterface,
    provider?: TerraformProvider,
  ) {
    checkIsValidId(id);

    super(scope, id);

    this.init(id, version, serverProps, provider);
  }

  private init(
    id: string,
    version: ApacheVersion,
    serverProps: ServerPropsInterface,
    provider?: TerraformProvider,
  ): void {
    const deployWith: ProviderType = serverProps.providerType;

    const imagesForProvider: Record<ApacheVersion, string> =
      this.supportedApacheImagesMap[deployWith];
    const imageIdentifier: string = imagesForProvider[version];

    validateImageExists(imageIdentifier);

    SingletonProviderFactory.getProvider(deployWith, this, provider);
    const deployStrategy: IDeployStrategy =
      ProviderDeployStrategyFactory.getProviderDeployStrategy(deployWith);
    this.createdApacheServer = this.deploy(
      deployStrategy,
      id,
      serverProps,
      imageIdentifier,
    );
  }

  protected abstract deploy(
    strategy: IDeployStrategy,
    id: string,
    props: ServerPropsInterface,
    imageIdentifier: string,
  ): Construct;

  protected getAdditionalProps(
    providerType: ProviderType,
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    switch (providerType) {
      case ProviderType.DOCKER:
        return this.getDockerProps(imageIdentifier);
      case ProviderType.AWS:
        return this.getAWSProps(imageIdentifier);
      default:
        return {};
    }
  }

  protected abstract getDockerProps(
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface;

  protected abstract getAWSProps(
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface;
}
