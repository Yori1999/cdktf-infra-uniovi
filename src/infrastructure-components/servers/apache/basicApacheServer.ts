import { ApacheServerBase } from "./apacheServerBase";
import {
  InternalMachineComponentPropsInterface,
  ServerPropsInterface,
} from "../../../props/props";
import { IDeployStrategy } from "../../../providers/providerDeployStrategy/deployStrategy";
import { supportedApacheImages } from "../../../supported-images/supportedServerImages";

export class ApacheServer extends ApacheServerBase {
  protected get supportedApacheImagesMap(): Record<
    string,
    Record<string, string>
  > {
    return supportedApacheImages;
  }

  protected deploy(
    strategy: IDeployStrategy,
    id: string,
    props: ServerPropsInterface,
    imageIdentifier: string,
  ): void {
    strategy.deployBasicServer(
      this,
      id,
      props,
      this.getAdditionalProps(props.providerType, imageIdentifier),
    );
  }

  protected getDockerProps(
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    return {
      dockerProps: {
        imageName: imageIdentifier,
      },
    };
  }

  protected getAWSProps(
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    return {
      awsProps: {
        ami: imageIdentifier,
        customInitScriptPath: "apache-basic-server",
      },
    };
  }
}
