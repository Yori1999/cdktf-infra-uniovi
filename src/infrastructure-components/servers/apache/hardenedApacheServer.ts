import { ApacheServerBase } from "./apacheServerBase";
import {
  InternalMachineComponentPropsInterface,
  ServerPropsInterface,
} from "../../../props/props";
import { IDeployStrategy } from "../../../providers/providerDeployStrategy/deployStrategy";
import { supportedHardenedApacheImages } from "../../../supported-images/supportedServerImages";

export class HardenedApacheServer extends ApacheServerBase {
  protected get supportedApacheImagesMap(): Record<
    string,
    Record<string, string>
  > {
    return supportedHardenedApacheImages;
  }

  protected deploy(
    strategy: IDeployStrategy,
    id: string,
    props: ServerPropsInterface,
    imageIdentifier: string,
  ): void {
    strategy.deployHardenedServer(
      this,
      id,
      props,
      this.getAdditionalProps(props.providerType, imageIdentifier),
    );
  }

  protected getDockerProps(
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    return {
      dockerProps: {
        imageName: imageIdentifier,
        // In this case, we'll be using a custom hardened image, so no need to curate a custom Dockerfile
      },
    };
  }

  protected getAWSProps(
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    return {
      awsProps: {
        ami: imageIdentifier,
        customInitScriptPath: "apache-hardened",
      },
    };
  }
}
