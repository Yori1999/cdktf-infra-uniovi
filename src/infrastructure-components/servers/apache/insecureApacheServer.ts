import { Construct } from "constructs";
import { ApacheServerBase } from "./apacheServerBase";
import {
  InternalMachineComponentPropsInterface,
  ServerPropsInterface,
} from "../../../props/props";
import { IDeployStrategy } from "../../../providers/providerDeployStrategy/deployStrategy";
import { supportedApacheImages } from "../../../supported-images/supportedServerImages";

export class InsecureApacheServer extends ApacheServerBase {
  protected get supportedApacheImagesMap(): Record<
    string,
    Record<string, string>
  > {
    return supportedApacheImages;
  }

  protected deploy(
    strategy: IDeployStrategy,
    id: string,
    props: ServerPropsInterface,
    imageIdentifier: string,
  ): Construct {
    return strategy.deployInsecureServer(
      this,
      id,
      props,
      this.getAdditionalProps(props.providerType, imageIdentifier),
    );
  }

  protected getDockerProps(
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    return {
      dockerProps: {
        imageName: imageIdentifier,
        customImageName: "apache-insecure",
        dockerfilePath: "apache-insecure",
      },
    };
  }

  protected getAWSProps(
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    return {
      awsProps: {
        ami: imageIdentifier,
        customInitScriptPath: "apache-insecure",
      },
    };
  }
}
