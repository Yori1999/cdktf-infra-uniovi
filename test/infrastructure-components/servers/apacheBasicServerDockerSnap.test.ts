import { App, TerraformStack, Testing } from "cdktf";
import { ApacheServer } from "../../../src/infrastructure-components/servers/apache/basicApacheServer";
import { ProviderType } from "../../../src/providers/providerType";
import { ApacheVersion } from "../../../src/supported-images/supportedServerImages";

function createTestStack(
  version: ApacheVersion,
  useVolume: boolean,
): TerraformStack {
  const app = new App();
  const stack = new TerraformStack(app, `test-${version}`);

  new ApacheServer(stack, "Test", version, {
    providerType: ProviderType.DOCKER,
    dockerProps: {
      useVolume: useVolume,
    },
  });

  return stack;
}

const apacheVersions = [
  { version: ApacheVersion.LATEST, expectedImage: "httpd:latest" },
  { version: ApacheVersion.APACHE_DEBIAN, expectedImage: "httpd:bookworm" },
  { version: ApacheVersion.APACHE_ALPINE, expectedImage: "httpd:alpine" },
];

describe("Apache Basic Server Docker container creation", () => {
  test.each(apacheVersions)(
    "creates a Docker container for %o",
    ({ version, expectedImage }) => {
      const stack = createTestStack(version, false);
      const synth = Testing.synth(stack);
      const parsed = JSON.parse(synth);

      expect(parsed).toMatchSnapshot();
      expect(parsed.resource.docker_container).toBeDefined();

      const image = parsed.resource.docker_image["Test_Test-image_5197E2DB"];
      expect(image.name).toBe(expectedImage);
      expect(
        parsed.resource.docker_container["Test_Test-container_72A6054C"].ports,
      ).toHaveLength(2);
    },
  );

  test("creates a Docker container with volume", () => {
    const stack = createTestStack(ApacheVersion.LATEST, true);
    const synth = Testing.synth(stack);
    const parsed = JSON.parse(synth);

    expect(parsed).toMatchSnapshot();

    const container =
      parsed.resource.docker_container["Test_Test-container_72A6054C"];
    expect(container.volumes).toHaveLength(1);
  });
});
