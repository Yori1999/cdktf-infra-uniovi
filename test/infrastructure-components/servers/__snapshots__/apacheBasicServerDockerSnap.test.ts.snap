// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Apache Basic Server Docker container creation creates a Docker container for { version: 'apacheAlpine', expectedImage: 'httpd:alpine' } 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
        "ports": [
          {
            "external": 8080,
            "internal": 80,
          },
          {
            "external": 8443,
            "internal": 443,
          },
        ],
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "httpd:alpine",
      },
    },
  },
}
`;

exports[`Apache Basic Server Docker container creation creates a Docker container for { version: 'apacheDebian', expectedImage: 'httpd:bookworm' } 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
        "ports": [
          {
            "external": 8080,
            "internal": 80,
          },
          {
            "external": 8443,
            "internal": 443,
          },
        ],
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "httpd:bookworm",
      },
    },
  },
}
`;

exports[`Apache Basic Server Docker container creation creates a Docker container for { version: 'latest', expectedImage: 'httpd:latest' } 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "provider": {
    "docker": [
      {},
    ],
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
        "ports": [
          {
            "external": 8080,
            "internal": 80,
          },
          {
            "external": 8443,
            "internal": 443,
          },
        ],
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "httpd:latest",
      },
    },
  },
  "terraform": {
    "required_providers": {
      "docker": {
        "source": "kreuzwerker/docker",
        "version": "3.0.2",
      },
    },
  },
}
`;

exports[`Apache Basic Server Docker container creation creates a Docker container with volume 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
        "ports": [
          {
            "external": 8080,
            "internal": 80,
          },
          {
            "external": 8443,
            "internal": 443,
          },
        ],
        "volumes": [
          {
            "container_path": "/mnt/data",
            "volume_name": "\${docker_volume.Test_Test-volume_F1ABD0A9.name}",
          },
        ],
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "httpd:latest",
      },
    },
    "docker_volume": {
      "Test_Test-volume_F1ABD0A9": {
        "name": "Test-data",
      },
    },
  },
}
`;
